envoy-filters/mgw-source/filters/http/mgw/mgw.cc

```diff
@@ -0,0 +1,136 @@
+#include "mgw-source/filters/http/mgw/mgw.h"
+
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/buffer/buffer_impl.h"
+#include "common/http/utility.h"
+#include "common/router/config_impl.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace MGW {
+
+Filter::Filter() {}
+
+// Http::StreamFilterBase
+void Filter::onDestroy() { ENVOY_LOG(debug, "filter destroyed"); }
+
+// Http::StreamEncoderFilter
+Http::FilterHeadersStatus Filter::encode100ContinueHeaders(Http::ResponseHeaderMap&) {
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus Filter::encodeHeaders(Http::ResponseHeaderMap&, bool) {
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus Filter::encodeData(Buffer::Instance&, bool) {
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus Filter::encodeTrailers(Http::ResponseTrailerMap&) {
+  return Http::FilterTrailersStatus::Continue;
+}
+
+Http::FilterMetadataStatus Filter::encodeMetadata(Http::MetadataMap&) {
+  return Http::FilterMetadataStatus::Continue;
+}
+
+void Filter::setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callbacks) {
+  res_callbacks_ = &callbacks;
+}
+
+// Http::StreamDecoderFilter
+Http::FilterTrailersStatus Filter::decodeTrailers(Http::RequestTrailerMap&) {
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void Filter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+  req_callbacks_ = &callbacks;
+}
+
+Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers, bool end_stream) {
+  ENVOY_LOG(trace, "decodeHeaders with end_stream = {}", end_stream);
+  req_headers_ = &headers;
+  // read metadata and check we need to modify the request body
+  set_body_ = readMetadata(&req_callbacks_->streamInfo().dynamicMetadata());
+  //if we do not modify the body we continue to other filters
+  if (!set_body_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  // if stream is not ended, stop continueing to other filters and buffer. 
```

Comment:
Let's drain the buffer since we modify the payload anyway. to avoid complications like buffer overflows.